CRM/Utils/System.php

```diff
@@ -1861,4 +1862,20 @@ public static function createDefaultCrudLink($crudLinkSpec) {
     return NULL;
   }
 
+  /**
+   * Return an HTTP Response with appropriate content and status code set.
+   * @param array $responseData
+   */
+  public static function sendResponse($responseData) {
```

Comment:
1. Right, so this is capturing the main goal of extracting that logic and creating a `CRM_Utils_System::sendResponse(...)` function.

2. I think this is too conservative in refactoring-technique -- i.e. it uses (by happenstance) the "HTTP-response data-model" from `AssetBuilder` rather than an "HTTP-response data-model" that's designed for general HTTP expectations. In particular, note that it supports one setting for one HTTP header (`mimeType` / `Content-Type`) -- but HTTP end-points may need to send many other headers. There are also edge-casey bits in handling the HTTP data (like distinguishing the status-code and the status-reason; like case-sensitivity in header names) which we haven't specified. [PSR-7](https://www.php-fig.org/psr/psr-7/) is not perfect, but it does address the edgy bits, and it's standard enough that we may eventually get some leverage out of it.

    Consider this signature:

    ```php
    function sendResponse(\Psr\Http\Message\ResponseInterface $response)
    ```

    and usage:

    ```php
    CRM_Utils_System::sendResponse(
      new \GuzzleHttp\Psr7\Response(200, ['Content-Type' => 'text/csv'], "ab,cd\n12,34\n")
    );
    ```

3. IMHO, the entire implementation of the `sendResponse()` function should live in `CRM_Utils_System_Base` so that it can be overriden on a per-CMS basis.

4. As long as the interface is based on PSR-7, I don't think there's much risk that we're going to be changing the signature/inputs.

Reply:
@totten i have updated now as per your sample code and i think it looks pretty sensible now